Initial prototype of org module

rjerrems opened the pull request:
Hi Guys,

Would be great to get some feedback on the structure & inclusions of the module. Once reviewed and I have made any updates, I'll finalize documentation / tests.

Cheers,
Rohan

rjerrems commented on examples/simple/outputs.tf:
```diff
@@ -14,12 +14,15 @@
-  description = "The name of the bucket to create."
-  type        = string
+output "org_terraform_sa_email" {
+  value = module.seed_bootstrap.org_terraform_sa_email
```
Agreed - I'll fix this

aaron-lane commented on variables.tf:
```diff
@@ -14,10 +14,94 @@
+  Optional variables
+*******************************************/
+
+variable "credentials_file_path" {
```
This can be removed: our standard is to defer to the supported credentials environment variables rather than using an input variable.

rjerrems replied:
Yep, I realized this is a legacy from a previous version and is no longer required.

aaron-lane commented on README.md:
```diff
@@ -15,68 +11,95 @@ module "bootstrap" {
+### Credentials
 
-- Storage Admin: `roles/storage.admin`
+If you are not customizing the credentials used in the variable `credentials_file_path`, please ensure you run `gcloud auth application-default login` prior to running your code to ensure the module can retreive credentials.
```
This statement should be removed along with the associated variable.

rjerrems replied:
Agreed, will pull it out.

aaron-lane commented on versions.tf:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright 2018 Google LLC
+ * Copyright 2019 Google LLC
```
The version should be pinned to `~> 0.12.6` to include support of `for_each` and to restrict support of 0.13.

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,151 @@
+}
+
+resource "google_project_service" "seed_project_api" {
+  count                      = length(local.activate_apis)
```
This should use `for_each` to allow for nondestructive changes in the APIs list.

```suggestion
  for_each = toset(local.activate_apis)
  
```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,151 @@
+resource "google_project_service" "seed_project_api" {
+  count                      = length(local.activate_apis)
+  project                    = google_project.seed_project.id
+  service                    = local.activate_apis[count.index]
```
```suggestion
  service                    = local.activate_apis[each.value]
```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,151 @@
+ ***********************************************/
+
+resource "google_organization_iam_member" "org_admins_group" {
+  count  = length(var.org_admins_org_iam_permissions)
```
```suggestion
  for_each = toset(var.org_admins_org_iam_permissions)
  
```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,151 @@
+resource "google_organization_iam_member" "org_admins_group" {
+  count  = length(var.org_admins_org_iam_permissions)
+  org_id = var.organization_id
+  role   = var.org_admins_org_iam_permissions[count.index]
```
```suggestion
  role   = var.org_admins_org_iam_permissions[each.value]
```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,151 @@
+ ***********************************************/
+
+resource "google_organization_iam_member" "tf_sa_org_perms" {
+  count  = length(var.sa_org_iam_permissions)
```
```suggestion
for_each = toset(var.sa_org_iam_permissions)

```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,151 @@
+resource "google_organization_iam_member" "tf_sa_org_perms" {
+  count  = length(var.sa_org_iam_permissions)
+  org_id = var.organization_id
+  role   = var.sa_org_iam_permissions[count.index]
```
```suggestion
  role   = var.sa_org_iam_permissions[each.value]
```

rjerrems commented:
Thanks @aaron-lane - I have updated in line with your feedback. Let me know if there is anything else you would like changed, I will start working on tests.

rjerrems commented:
@aaron-lane @marine675 - I have added some basic tests for main module and submodule. If you could take a look and let me know if needs more work that would be great.

marine675 commented on modules/cloudbuild/main.tf:
```diff
@@ -0,0 +1,249 @@
+  cloudbuild_apis             = ["cloudbuild.googleapis.com", "sourcerepo.googleapis.com", "cloudkms.googleapis.com"]
+  impersonation_enabled_count = var.sa_enable_impersonation == true ? 1 : 0
+  activate_apis               = distinct(concat(var.activate_apis, local.cloudbuild_apis))
+  csr_repos                   = var.cloud_source_repos
```
this local variable is not used

rjerrems replied:
Good catch, will remove

rjerrems commented on variables.tf:
```diff
@@ -14,10 +14,102 @@
+  default = [
+    "roles/billing.user",
+    "roles/resourcemanager.organizationAdmin",
+    "roles/resourcemanager.projectCreator"
```
True - I'll remove

rjerrems commented on modules/cloudbuild/cloudbuild_builder/README.markdown:
```diff
@@ -0,0 +1,14 @@
+# [Terraform](https://www.terraform.io/docs) cloud builder
```
Cool will do

aaron-lane commented on build/int.cloudbuild.yaml:
```diff
@@ -21,18 +21,21 @@ steps:
   - 'TF_VAR_org_id=$_ORG_ID'
   - 'TF_VAR_folder_id=$_FOLDER_ID'
   - 'TF_VAR_billing_account=$_BILLING_ACCOUNT'
+  - 'TF_VAR_group_org_admins=$_GROUP_ORG_ADMINS'
```
```suggestion
  - 'TF_VAR_group_org_admins=test-gcp-org-admins@test.infra.cft.tips'
```

aaron-lane commented on build/int.cloudbuild.yaml:
```diff
@@ -21,18 +21,21 @@ steps:
   - 'TF_VAR_folder_id=$_FOLDER_ID'
   - 'TF_VAR_billing_account=$_BILLING_ACCOUNT'
+  - 'TF_VAR_group_org_admins=$_GROUP_ORG_ADMINS'
+  - 'TF_VAR_group_billing_admins=$_GROUP_BILLING_ADMINS'
```
```suggestion
  - 'TF_VAR_group_billing_admins=test-gcp-billing-admins@test.infra.cft.tips'
```

aaron-lane commented on test/fixtures/cloudbuild_enabled/variables.tf:
```diff
@@ -0,0 +1,56 @@
+}
+
+variable "default_region" {
+  description = "Default region to create resources where applicable."
```
```suggestion
  description = "Default region to create resources where applicable."
  default     = "us-central-1"
```

aaron-lane commented on README.md:
```diff
@@ -1,10 +1,6 @@
-- Create a GCS bucket with the provided name
+The purpose of this module is to help bootstrap a GCP organization, creating all the required GCP resources & permissions to start using the Cloud Foundation Toolkit (CFT). For users who want to use Cloud Build & Cloud Source Repos for foundations code, there is also a submodule to help bootstrap all the required resources to do this.
 
 ## Usage
```
We need a caveat section which explains that multiple applies may be required due to the dependency on the unmanaged Cloud Build service account.

rjerrems replied:
I believe I have resolved that issue now, I have been getting consistent results both locally and last couple of integration tests.

aaron-lane commented on examples/cloudbuild_enabled/outputs.tf:
```diff
@@ -0,0 +1,59 @@
+ * limitations under the License.
+ */
+
+output "group_org_admins" {
```
The example should not need to output variables; this can be handled in the test fixture.

aaron-lane replied:
Additionally, outputs should have descriptions.

rjerrems replied:
no worries, I have fixed in latest commit.

aaron-lane commented on examples/cloudbuild_enabled/variables.tf:
```diff
@@ -0,0 +1,41 @@
+ * limitations under the License.
+ */
+
+variable "organization_id" {
```
Variables must be typed for clarity.

rjerrems replied:
No problem - will update

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,141 @@
+  source                      = "terraform-google-modules/project-factory/google"
+  version                     = "~> 5.0"
+  name                        = local.seed_project_id
+  random_project_id           = "true"
```
```suggestion
  random_project_id           = true
```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,141 @@
+  version                     = "~> 5.0"
+  name                        = local.seed_project_id
+  random_project_id           = "true"
+  disable_services_on_destroy = "false"
```
```suggestion
  disable_services_on_destroy = false
```

aaron-lane commented on main.tf:
```diff
@@ -14,11 +14,141 @@
+  project       = module.seed_project.project_id
+  name          = format("%s-%s-%s", var.project_prefix, "tfstate", random_id.suffix.hex)
+  location      = var.default_region
+  force_destroy = true
```
This seems dangerous. I think users should actively ensure that all Terraform state has been removed from the state bucket before destroying it.

rjerrems replied:
Yeah true, this was a relic of me creating / blowing up environments end to end. I'll remove this.

aaron-lane commented on modules/cloudbuild/cloudbuild_builder/README.md:
```diff
@@ -0,0 +1,14 @@
+# [Terraform](https://www.terraform.io/docs) cloud builder
```
I suggest moving the link from this header to the body of the next section.

rjerrems replied:
I have tidied this up, was derived from the community terraform builder.

aaron-lane commented on modules/cloudbuild/main.tf:
```diff
@@ -0,0 +1,280 @@
+  source                      = "terraform-google-modules/project-factory/google"
+  version                     = "~> 5.0"
+  name                        = local.cloudbuild_project_id
+  random_project_id           = "true"
```
```suggestion
  random_project_id           = true
```

aaron-lane commented on modules/cloudbuild/main.tf:
```diff
@@ -0,0 +1,280 @@
+  version                     = "~> 5.0"
+  name                        = local.cloudbuild_project_id
+  random_project_id           = "true"
+  disable_services_on_destroy = "false"
```
```suggestion
  disable_services_on_destroy = false
```

aaron-lane commented on modules/cloudbuild/main.tf:
```diff
@@ -0,0 +1,280 @@
+  activate_apis               = local.activate_apis
+}
+
+resource "google_project_service" "cloudbuild_apis" {
```
It should be possible to move these in to the Project Factory module and enforce dependencies through the `project_id` output as it blocks on the service resources.

rjerrems replied:
Yeah agree, although my experience has been that there are occasional race conditions using this approach. So it is a trade off between stability vs how concise / clean the code is. I can switch back but will likely cause intermittent failures in tests and also in wild?

aaron-lane replied:
The two approaches should be functionally equivalent. I would like to understand why if they aren't but we don't need to block on this.

aaron-lane commented on examples/cloudbuild_enabled/cloudbuild-tf-plan.yaml:
```diff
@@ -12,10 +12,20 @@
+- name: gcr.io/$PROJECT_ID/terraform
+  args:
+  - init
+  - -input=false
```
We should include `terraform validate` before `terraform plan`.

rjerrems replied:
Yep makes sense, I'll add that

aaron-lane commented on test/fixtures/cloudbuild_enabled/outputs.tf:
```diff
@@ -0,0 +1,55 @@
+ * limitations under the License.
+ */
+
+output "group_org_admins" {
```
If we need to output variables for test inputs then it should be done here rather than in the example.

rjerrems replied:
ack, will fix

aaron-lane commented on test/fixtures/simple/variables.tf:
```diff
@@ -0,0 +1,56 @@
+}
+
+variable "default_region" {
+  description = "Default region to create resources where applicable."
```
```suggestion
  description = "Default region to create resources where applicable."
  default     = "us-central-1"
```

aaron-lane commented on test/integration/cloudbuild_enabled/controls/gcp.rb:
```diff
@@ -0,0 +1,107 @@
+    its('has_user_managed_keys?') {should cmp false }
+  end
+
+  default_apis.each do |api|
```
We should avoid dynamically defining test cases when possible.

```ruby
describe google_project_services(project: attribute("seed_project_id")).where(state: 'ENABLED') do
  its('names') { should match_array default_apis }
end
```

rjerrems replied:
I am going to leave this as is, because using `google_project_services` adds a lot of unnecessary code as the names are in the format of `projects/<project_number>/services/<API>.googleapis.com` let me know if you think it has to change.

rjerrems commented on test/setup/make_source.sh:
```diff
@@ -16,8 +16,13 @@
 
-project_id=$(terraform output project_id)
-echo "export TF_VAR_project_id='$project_id'" >> ../source.sh
+TF_VARS="project_id org_id folder_id billing_account group_org_admins group_billing_admins default_region org_project_creators"
```
This does not seem to be true locally for me, using `0.6.0` - should I be using a newer image?

aaron-lane replied:
The latest image should be able to read the outputs of the `setup` configuration and automatically use them as variable inputs to the test fixtures without the need for `make_source.sh`. That does not appear to be working?

rjerrems replied:
Looks like you were right, I may have tested agains't an older image. I have removed. Thanks!

rjerrems commented:
@aaron-lane thanks again for review. I believe I have addressed all the feedback, except items I have left unresolved. Let me know what you think when you have a chance.
